// turbo-cli/template/community/models/Comment.js
/*
  This is a schema based on the NeDB local database which follows the
  MongoDB API (https://www.npmjs.com/package/nedb). The 'camo' library
  is an ORM for the NeDB implementation.

  Eventually, this should be replaced by a MONGOOSE schema when used in
  conjunction with Mongo DB. This would happen in the case of a
  developer taking over the project.
*/


// https://github.com/scottwrobinson/camo
const Document = require('vertex-camo').Document
const props = {
  profile: {type:String, default:''},
	title: {type:String, default:'', display:true},
  slug: {type:String, default:'', immutable:true},
	text: {type:String, default:''},
	url: {type:String, default:''},
	domain: {type:String, default:'', immutable:true}, // www.nypost.com
	attachment: {type:String, default:''},
	image: {type:String, default:''},
	thread: {type:String, default:'', immutable:true}, // id number of post comment refers to
	isInitial: {type:String, default:'yes', immutable:true},
	votes: {type:Object, default:{up:[], down:[], score:0}, immutable:true},
	numReplies: {type:Number, default:0, immutable:true},
	updated: {type:Date, default:Date.now, immutable:true},
	dateString: {type:String, default:'', immutable:true},
	timestamp: {type:Date, default:Date.now, immutable:true},
}

class Comment extends Document {
	constructor(){
		super()
		this.schema(props)
	}

	static get resourceName(){
		return 'comment'
	}

	static collectionName(){
			return 'comments'
	}

}

module.exports = Comment
