// turbo-cli/dist/template/community/controllers/CommentController.js
const vertex = require('vertex360')({site_id: process.env.TURBO_APP_ID})
const Comment = require('../models/Comment')
const Controller = vertex.Controller

class CommentController extends Controller {
	constructor(){
		super(Comment, process.env)
	}

	get(params) {
		return new Promise((resolve, reject) => {
			Controller.checkCollectionDB(Comment.collectionName())
			.then(data => {
				return Comment.find(params, Controller.parseFilters(params))
			})
			.then(comments => {
				resolve(Comment.convertToJson(comments))
			})
			.catch(err => {
				reject(err)
			})
		})
	}

	getById(id) {
		return new Promise((resolve, reject) => {
			Controller.checkCollectionDB(Comment.collectionName())
			.then(data => {
				return Comment.findById(id)
			})
			.then(comment => {
				if (comment == null){
					throw new Error(Comment.resourceName + ' ' + id + ' not found.')
					return
				}

				resolve(comment.summary())
			})
			.catch(err => {
				reject(new Error(Comment.resourceName + ' ' + id + ' not found.'))
			})
		})
	}

	post(body) {
		return new Promise((resolve, reject) => {
			let payload = null
			body['dateString'] = vertex.utils.formattedDate()

			Comment.create(body)
			.then(comment => {
				payload = comment.summary()
				return (process.env.TURBO_ENV=='dev') ? null : Controller.syncCollection(Comment.collectionName())
			})
			.then(data => {
				resolve(payload)
			})
			.catch(err => {
				reject(err)
			})
		})
	}

	put(id, params) {
		return new Promise((resolve, reject) => {
			let payload = null
			Comment.findByIdAndUpdate(id, params, {new:true})
			.then(comment => {
				payload = comment.summary()
				return (process.env.TURBO_ENV=='dev') ? null : Controller.syncCollection(Comment.collectionName())
			})
			.then(data => {
				resolve(payload)
			})
			.catch(err => {
				reject(err)
			})
		})
	}

	delete(id) {
		return new Promise((resolve, reject) => {
			Comment.findByIdAndRemove(id)
			.then(() => {
				return (process.env.TURBO_ENV=='dev') ? null : Controller.syncCollection(Comment.collectionName())
			})
			.then(data => {
				resolve()
			})
			.catch(err => {
				reject(err)
			})
		})
	}
}

module.exports = CommentController
