// turbo-cli/dist/extend/graphql/gql/schema.js
const { GraphQLObjectType, GraphQLID, GraphQLString, GraphQLSchema, GraphQLList, GraphQLNonNull } = require('graphql')
const Profile = require('../models/Profile')
const types = require('./types')

const RootQuery = new GraphQLObjectType({
  name: 'RootQueryType',
  fields: () => ({ // set up individual queries in here:
    profiles: {
			type: new GraphQLList(types.profile),
      args: {
        email: {type: GraphQLString}
      },
      resolve: (parent, args) => Profile.find(args)
		},
    profile: {
      type: types.profile,
      args: {
        id: {type: GraphQLID}
      },
      resolve: (parent, args) => Profile.findById(args.id)
    }
  })
})

const Mutation = new GraphQLObjectType({
	name: 'Mutation',
	fields: () => ({
    createProfile: {
      type: types.profile,
      args: {
        email: {type: new GraphQLNonNull(GraphQLString)}
      },
      resolve: (parent, args) => {
        return Profile.create(args)
      }
    }
	})
})


module.exports = new GraphQLSchema({
  query: RootQuery,
  mutation: Mutation
})
